### PR TITLE
lightning: add retry and early terminate checking empty table (#31798)

### LINKED ISSUE #31797: lightning get stuck when check table empty failed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Run tidb-lightning with 1k tables, there is a low probability that tidb-lightning will stuck.

From the go routine, I found it was stuck at https://github.com/pingcap/tidb/blob/7fd298d916c30ed0060db3581be58b448325819e/br/pkg/lightning/restore/check_info.go#L1132

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #31798 to release-5.4
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/33148
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/33148:release-5.4-a702ef1b7739
```

---

### What problem does this PR solve?

Issue Number: close #31797

Problem Summary:

### What is changed and how it works?
- Add check `ctx.Done` when send msg to chan
- Add retry for exec `select` to avoid tidb failure. This can alleviate #31318
- Avoid output misleading log in tidb backend

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Add retry to avoid precheck failure when query execution timeout
```

### CODE REVIEW

File: br/pkg/lightning/restore/check_info_test.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ import (
 	"context"
 	"database/sql"
 	"fmt"
+	"github.com/stretchr/testify/require"
```
Comment (gozssky) [inline br/pkg/lightning/restore/check_info_test.go:21]: Imports are not sorted.

File: br/pkg/lightning/restore/meta_manager.go [hunk 2]
```diff
@@ -206,7 +206,11 @@ func (m *dbTableMetaMgr) AllocTableRowIDs(ctx context.Context, rawRowIDMax int64
 			}
 
 			if status == metaStatusChecksuming {
+<<<<<<< HEAD
 				return errors.New("target table is calculating checksum, please wait unit the checksum is finished and try again.")
+=======
+				return common.ErrAllocTableRowIDs.GenWithStack("Target table is calculating checksum. Please wait until the checksum is finished and try again.")
+>>>>>>> a702ef1b7... lightning: add retry and early terminate checking empty table (#31798)
```
Comment (gozssky) [inline br/pkg/lightning/restore/meta_manager.go:213]: Merge conflicts.

**Review by glorv — APPROVED** [review]

**Review by gozssky — APPROVED** [review]

**Review by gozssky — APPROVED** [review]

**Pull request merged**